pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+	//   to cleanup jobs with old labels
+	jm.enqueueController(curr)
+}
+
+// syncCronJob reconciles a CronJob with a list of any Jobs that it created.
+// All known jobs created by "cj" should be included in "js".
+// The current time is passed in to facilitate testing.
+// It returns a copy of the CronJob that is to be used by other functions
+// that mutates the object
+func (jm *ControllerV2) syncCronJob(
+	cj *batchv1beta1.CronJob,
+	js []batchv1.Job) (*batchv1beta1.CronJob, *time.Duration, error) {
+
+	cj = cj.DeepCopy()
+	now := jm.now()
+
+	childrenJobs := make(map[types.UID]bool)
+	for _, j := range js {
+		childrenJobs[j.ObjectMeta.UID] = true
+		found := inActiveList(*cj, j.ObjectMeta.UID)
+		if !found && !IsJobFinished(&j) {
+			jm.recorder.Eventf(cj, corev1.EventTypeWarning, "UnexpectedJob", "Saw a job that the controller did not create or forgot: %s", j.Name)
+			// We found an unfinished job that has us as the parent, but it is not in our Active list.
+			// This could happen if we crashed right after creating the Job and before updating the status,
+			// or if our jobs list is newer than our cj status after a relist, or if someone intentionally created
+			// a job that they wanted us to adopt.
+		} else if found && IsJobFinished(&j) {
+			_, status := getFinishedStatus(&j)
+			deleteFromActiveList(cj, j.ObjectMeta.UID)
+			jm.recorder.Eventf(cj, corev1.EventTypeNormal, "SawCompletedJob", "Saw completed job: %s, status: %v", j.Name, status)
+		}
+	}
+
+	// Remove any job reference from the active list if the corresponding job does not exist any more.
+	// Otherwise, the cronjob may be stuck in active mode forever even though there is no matching
+	// job running.
+	for _, j := range cj.Status.Active {
+		_, found := childrenJobs[j.UID]
+		if found {
+			continue
+		}
+		// Explicitly try to get the job from api-server to avoid a slow watch not able to update
+		// the job lister on time, giving an unwanted miss
+		_, err := jm.jobControl.GetJob(j.Namespace, j.Name)
+		switch {
+		case err != nil && errors.IsNotFound(err):
+			// The job is actually missing, delete from active list and schedule a new one if within
+			// deadline
+			jm.recorder.Eventf(cj, corev1.EventTypeNormal, "MissingJob", "Active job went missing: %v", j.Name)
+			deleteFromActiveList(cj, j.UID)
+		case err != nil:
+			return cj, nil, err
+		}
+		// the job is missing in the lister but found in api-server
+	}
+
+	updatedCJ, err := jm.cronJobControl.UpdateStatus(cj)
+	if err != nil {
+		klog.V(2).InfoS("Unable to update status for cronjob", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()), "resourceVersion", cj.ResourceVersion, "err", err)
+		return cj, nil, err
+	}
+	*cj = *updatedCJ
+
+	if cj.DeletionTimestamp != nil {
+		// The CronJob is being deleted.
+		// Don't do anything other than updating status.
+		return cj, nil, fmt.Errorf("cronjob %s/%s is being deleted", cj.Namespace, cj.Name)
```

Comment:
Why does this return an error?  Seems like we just want to return no error and no further action. Otherwise this will continually self-trigger, right?

Reply:
+1, no error here, I had the changes stashed from the last review, will push them, sorry about the slopiness!

Reply:
done